### PR TITLE
Fixes wrong number or read/written bytes for System.Device.I2c

### PR DESCRIPTION
**Pull request by RolandoMagico**
## Description
- Add code to take into account the start and length of SpanByte for read/written operations.

## Motivation and Context
- Fixes nanoframework/home#674.

## How Has This Been Tested?
Not yet tested

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/nanoframework/nf-interpreter?pullRequest=1819) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/nanoframework/nf-interpreter?pullRequest=1819) before we can accept your contribution.<br/><br/>:x: RolandoMagico [sign now](https://cla.dotnetfoundation.org/nanoframework/nf-interpreter?pullRequest=1819)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/nanoframework/nf-interpreter?pullRequest=1819) it.</sub>

**Comment by josesimoes** [comment]
Hey @RolandoMagico !

First of all: thank you for looking into this. 👍🏻 

Please use separate PRs to address each issue: one for the wrong link and another for fixing the SpanByte issue with I2C.

**Comment by josesimoes** [comment]
@RolandoMagico it would be awesome if you could replicate the fix to the STM32 implementation... 😉

File: targets/FreeRTOS_ESP32/ESP32_WROOM_32/nanoCLR/System.Device.I2c/sys_dev_i2c_native_System_Device_I2c_I2cDevice.cpp [hunk 1]
```diff
@@ -130,11 +132,14 @@ HRESULT Library_sys_dev_i2c_native_System_Device_I2c_I2cDevice::
...
             {
-                // grab the pointer to the array by getting the first element of the array
-                writeData = writeBuffer->GetFirstElement();
+                // Get the write offset, only the elements defined by the span must be written, not the whole array
+                writeOffset = writeSpanByte[SpanByte::FIELD___start].NumericByRef().s4;
 
-                // get the size of the buffer by reading the number of elements in the HeapBlock array
-                writeSize = writeBuffer->m_numOfElements;
+                // grab the pointer to the array by starting and the offset specified in the span
+                writeData = writeBuffer->GetElement(writeOffset)();
```
Comment (josesimoes) [inline targets/FreeRTOS_ESP32/ESP32_WROOM_32/nanoCLR/System.Device.I2c/sys_dev_i2c_native_System_Device_I2c_I2cDevice.cpp:139]: No need for the extra () and the end

File: targets/FreeRTOS_ESP32/ESP32_WROOM_32/nanoCLR/System.Device.I2c/sys_dev_i2c_native_System_Device_I2c_I2cDevice.cpp [hunk 2]
```diff
@@ -144,11 +149,14 @@ HRESULT Library_sys_dev_i2c_native_System_Device_I2c_I2cDevice::
             readBuffer = readSpanByte[SpanByte::FIELD___array].DereferenceArray();
             if (readBuffer != NULL)
             {
-                // grab the pointer to the array by getting the first element of the array
-                readData = readBuffer->GetFirstElement();
+                // Get the read offset, only the elements defined by the span must be read, not the whole array
+                readOffset = readSpanByte[SpanByte::FIELD___start].NumericByRef().s4;
+
+                // grab the pointer to the array by starting and the offset specified in the span
+                readData = readBuffer->GetElement(readOffset)();
```
Comment (josesimoes) [inline targets/FreeRTOS_ESP32/ESP32_WROOM_32/nanoCLR/System.Device.I2c/sys_dev_i2c_native_System_Device_I2c_I2cDevice.cpp:156]: No need for the extra () at the end of the line

File: targets/FreeRTOS_ESP32/ESP32_WROOM_32/nanoCLR/System.Device.I2c/sys_dev_i2c_native_System_Device_I2c_I2cDevice.cpp [hunk 3]
```diff
@@ -144,11 +149,14 @@ HRESULT Library_sys_dev_i2c_native_System_Device_I2c_I2cDevice::
...
 
-                // get the size of the buffer by reading the number of elements in the HeapBlock array
-                readSize = readBuffer->m_numOfElements;
+                // use the span length as write size, only the elements defined by the span must be read
```
Comment (josesimoes) [inline targets/FreeRTOS_ESP32/ESP32_WROOM_32/nanoCLR/System.Device.I2c/sys_dev_i2c_native_System_Device_I2c_I2cDevice.cpp:158]: Comment should be "as READ size" not write. Possibly from copy/paste. 😉

**Review by josesimoes — CHANGES_REQUESTED** [review]

**Comment by josesimoes** [comment]
Ignoring build failure for ChibiOS based targets.

**Review by josesimoes — APPROVED** [review]
LGTM!

**Pull request merged**